File: src/renderer/db/cedict.ts
import XRegExp from "xregexp";

import db from "./data";

const baseQuery = `
SELECT * FROM vocab
INNER JOIN token ON token.entry = vocab.simplified
ORDER BY frequency DESC`;
const chineseRegex = XRegExp("\\p{Han}");

interface ICedict {
    id: number;
    simplified: string;
    traditional: string;
    pinyin: string;
    english: string;
    frequency: number;
}

interface ICedictMap {
    [key: string]: ICedict;
}

function _searchOne(s: string, col: string): Promise<ICedictMap> {
    return new Promise((resolve, reject) => {
        db.all(`
        ${baseQuery}
        WHERE ${col} LIKE ?`, [`%${s}%`], (err, rows) => { err ? reject(err) : resolve(rows.reduce((o, el) => {
            const {entry, ...cedict} = el;
            o[cedict.frequency.toString()] = cedict as ICedict;
            return o;
        }, {})); });
    });
}

export function searchChinese(s: string): Promise<ICedict[]> {
    return new Promise((resolve, reject) => {
        db.all(`
        ${baseQuery}
        WHERE simplified LIKE ? OR traditional LIKE ?`,
        [`%${s}%`, `%${s}%`], (err, rows) => { err ? reject(err) : resolve(rows.map((el) => {
            const {entry, ...cedict} = el;
            return cedict as ICedict;
        })); });
    });
}

export function searchChineseMatch(s: string): Promise<ICedict[]> {
    return new Promise((resolve, reject) => {
        db.all(`
        ${baseQuery}
        WHERE simplified = ? OR traditional = ?`,
        [s, s], (err, rows) => { err ? reject(err) : resolve(rows.map((el) => {
            const {entry, ...cedict} = el;
            return cedict as ICedict;
        })); });
    });
}

export async function searchEnglish(s: string): Promise<ICedict[]> {
    const map = await _searchOne(s, "english");
    return Object.keys(map).map((f) => parseFloat(f)).sort().map((f) => map[f]);
}

export async function searchPinyin(s: string): Promise<ICedict[]> {
    const map = await _searchOne(s, "pinyin");
    return Object.keys(map).map((f) => parseFloat(f)).sort().map((f) => map[f]);
}

export async function _searchNonChinese(s: string): Promise<ICedict[]> {
    const map = {
        ...await _searchOne(s, "english"),
        ...await _searchOne(s, "pinyin")
    };
    return Object.keys(map).map((f) => parseFloat(f)).sort().map((f) => map[f]);
}

export default function searchDict(s: string): Promise<ICedict[]> {
    return chineseRegex.test(s) ? searchChinese(s) : _searchNonChinese(s);
}

File: kotlin/src/main/kotlin/hanzileveljs/Cedict.kt
package hanzileveljs

class Cedict {
    data class CedictEntry internal constructor(
            val id: Int,
            val simplified: String,
            val traditional: String?,
            val pinyin: String,
            val english: String,
            val frequency: Float
    )

    private val baseQuery = """
        SELECT vocab.id AS id, simplified, traditional, pinyin, english, frequency FROM vocab
        INNER JOIN token ON token.entry = vocab.simplified
    """.trimIndent()

    private fun searchOne(s: String, col: String): Map<Float, CedictEntry> {
        val o = mutableMapOf<Float, CedictEntry>()

        Config.db.open().createQuery("""
            $baseQuery
            WHERE $col LIKE :s
            ORDER BY frequency DESC
        """.trimIndent())
                .addParameter("s", "%$s%")
                .executeAndFetch(CedictEntry::class.java)
                .forEach {
                    o[it.frequency] = it
                }

        return o.toMap()
    }

    private fun searchNonChinese(s: String): List<CedictEntry> {
        val output = searchOne(s, "pinyin").toMutableMap()
        output.putAll(searchOne(s, "english"))

        return output.values.sortedBy { it.frequency }
    }

    fun searchChinese(s: String): List<CedictEntry> {
        return Config.db.open().createQuery("""
            $baseQuery
            WHERE simplified LIKE :s OR traditional LIKE :s
            ORDER BY frequency DESC
        """.trimIndent())
                .addParameter("s", "%$s%")
                .executeAndFetch(CedictEntry::class.java)
    }

    fun searchChineseMatch(s: String): List<CedictEntry> {
        return Config.db.open().createQuery("""
            $baseQuery
            WHERE simplified = :s OR traditional = :s
            ORDER BY frequency DESC
        """.trimIndent())
                .addParameter("s", s)
                .executeAndFetch(CedictEntry::class.java)
    }

    fun searchPinyin(s: String): List<CedictEntry> {
        return Config.db.open().createQuery("""
            $baseQuery
            WHERE pinyin LIKE :s
            ORDER BY frequency DESC
        """.trimIndent())
                .addParameter("s", "%$s%")
                .executeAndFetch(CedictEntry::class.java)
    }

    fun searchEnglish(s: String): List<CedictEntry> {
        return Config.db.open().createQuery("""
            $baseQuery
            WHERE english LIKE :s
            ORDER BY frequency DESC
        """.trimIndent())
                .addParameter("s", "%$s%")
                .executeAndFetch(CedictEntry::class.java)
    }

    operator fun get(s: String): List<CedictEntry> = when(Regex("\\p{IsHan}").find(s)) {
        null -> searchNonChinese(s)
        else -> searchChinese(s)
    }
}